Bug: In v8, no-unused-vars says "defined but only used as type" on import with no escape hatch

### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

With code like:

```ts
import { SomeType } from "./someModule";

export const value = 1234 as typeof SomeType;
```

This code is flagged as having an unused variable in v8 with a new message, something like:

```
/home/jabaile/work/TypeScript/src/harness/client.ts
  53:5  warning  'PatternMatchKind' is defined but only used as a type. Allowed unused vars must match /^(_+$|_[^_])/u  @typescript-eslint/no-unused-vars
```

The TS repo canonically does not use type-only imports (we cannot use `verbatimModuleSyntax`), so it's not desirable to make this a type-only import. But, there's no way to opt out of this new detection for imports.

### Reproduction Repository Link

https://github.com/jakebailey/TypeScript

### Repro Steps

1) Clone https://github.com/jakebailey/TypeScript/tree/eslint-9
2) `npm ci`
3) `npx eslint src/harness/client.ts`

```
/home/jabaile/work/TypeScript/src/harness/client.ts
  53:5  warning  'PatternMatchKind' is defined but only used as a type. Allowed unused vars must match /^(_+$|_[^_])/u  @typescript-eslint/no-unused-vars
```

### Versions

<!-- Fill in the relevant versions and delete the others -->
| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `8.0.0-alpha.59` |
| `@typescript-eslint/parser`            | `8.0.0-alpha.59` |
| `@typescript-eslint/scope-manager`     | `8.0.0-alpha.59` |
| `@typescript-eslint/typescript-estree` | `8.0.0-alpha.59` |
| `@typescript-eslint/type-utils`        | `8.0.0-alpha.59` |
| `@typescript-eslint/utils`             | `8.0.0-alpha.59` |
| `TypeScript`                           | `5.4.4` |
| `ESLint`                               | `9.8.0` |
| `node`                                 | `22.5.1` |

--- a/packages/eslint-plugin/src/rules/no-unused-vars.ts
+++ b/packages/eslint-plugin/src/rules/no-unused-vars.ts
@@ -579,16 +579,6 @@ export default createRule<Options, MessageIds>({
         for (const unusedVar of unusedVars) {
           // Report the first declaration.
           if (unusedVar.defs.length > 0) {
-            const writeReferences = unusedVar.references.filter(
-              ref =>
-                ref.isWrite() &&
-                ref.from.variableScope === unusedVar.scope.variableScope,
-            );
-
-            const id = writeReferences.length
-              ? writeReferences[writeReferences.length - 1].identifier
-              : unusedVar.identifiers[0];
-
             const usedOnlyAsType = unusedVar.references.some(ref =>
               referenceContainsTypeQuery(ref.identifier),
             );
@@ -601,6 +591,16 @@ export default createRule<Options, MessageIds>({
             if (isImportUsedOnlyAsType) {
               continue;
             }
+
+            const writeReferences = unusedVar.references.filter(
+              ref =>
+                ref.isWrite() &&
+                ref.from.variableScope === unusedVar.scope.variableScope,
+            );
+
+            const id = writeReferences.length
+              ? writeReferences[writeReferences.length - 1].identifier
+              : unusedVar.identifiers[0];
 
             const messageId = usedOnlyAsType ? 'usedOnlyAsType' : 'unusedVar';
 

--- a/packages/eslint-plugin/src/rules/no-unused-vars.ts
+++ b/packages/eslint-plugin/src/rules/no-unused-vars.ts
@@ -593,6 +593,15 @@ export default createRule<Options, MessageIds>({
               referenceContainsTypeQuery(ref.identifier),
             );
 
+            const isImportUsedOnlyAsType =
+              usedOnlyAsType &&
+              unusedVar.defs.some(
+                def => def.type === DefinitionType.ImportBinding,
+              );
+            if (isImportUsedOnlyAsType) {
+              continue;
+            }
+
             const messageId = usedOnlyAsType ? 'usedOnlyAsType' : 'unusedVar';
 
             const { start } = id.loc;

--- a/packages/eslint-plugin/tests/rules/no-unused-vars/no-unused-vars.test.ts
+++ b/packages/eslint-plugin/tests/rules/no-unused-vars/no-unused-vars.test.ts
@@ -1173,6 +1173,16 @@ export type Foo = typeof foo;
 
 export const bar = (): Foo => foo;
     `,
+    `
+import { SomeType } from 'foo';
+
+export const value = 1234 as typeof SomeType;
+    `,
+    `
+import { foo } from 'foo';
+
+export type Bar = typeof foo;
+    `,
   ],
 
   invalid: [
@@ -2253,27 +2263,6 @@ export const x = _Foo;
           column: 15,
           endLine: 2,
           endColumn: 18,
-        },
-      ],
-    },
-    {
-      code: `
-        import { foo } from 'foo';
-
-        export type Bar = typeof foo;
-      `,
-      errors: [
-        {
-          messageId: 'usedOnlyAsType',
-          data: {
-            varName: 'foo',
-            action: 'defined',
-            additional: '',
-          },
-          line: 2,
-          column: 18,
-          endLine: 2,
-          endColumn: 21,
         },
       ],
     },

